### PR TITLE
Document the different stacks

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Documents how the different stacks are used, both in ARMv7/Aarch32 and
Aarch64.

Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

File: documentation/optee_design.md [hunk 1]
```diff
@@ -315,9 +315,105 @@ TA. This is solved by always setting the TA context in the CPU when
...
 than not that it's the same CPU that resumes execution in secure world.
 
-# 6. SVC
-Will be written soon.
-
+# 6. Stacks
+Different stacks are used during different stages. The stacks are:
+- Secure monitor stack (128 bytes), bound to the CPU. Only available if
+  OP-TEE is compiled with a secure monitor always the case if the target is
+  ARMv7 but never for ARMv8.
```
Comment (jbech-linaro) [inline documentation/optee_design.md]: We quite often get comments from others that we should be more precise with ARMv7/v8 and write ARMv7-A, ARMv8-A instead.

File: documentation/optee_design.md [hunk 2]
```diff
@@ -315,9 +315,105 @@ TA. This is solved by always setting the TA context in the CPU when
...
+  one state to another. Interrupts are always disabled when using this
+  stack, aborts are fatal when using the temp stack.
+- Abort stack (medium ~2KB), bound to the CPU. Used when trapping a data
+  or pre-fetch abort. Aborts from user space are never fatal the TA is only
+  killed. Aborts from kernel mode are used by the pager to do the demand
+  paging, if pager is disabled all kernel mode aborts are fatal.
+- Thread stack (large ~8KB), not bound to the CPU instead used by the current
+  thread/task. Interrupts are usually enabled when using this stack.
+
+*Notes for ARMv7/Aarch32:*
```
Comment (jbech-linaro) [inline documentation/optee_design.md]: s/Aarch32/AArch32/g

File: documentation/optee_design.md [hunk 3]
```diff
@@ -315,9 +315,105 @@ TA. This is solved by always setting the TA context in the CPU when
...
+
+| Stack  | Comment |
+|--------|---------|
+| Temp   | Assigned to `SP_SVC` during entry/exit, always assigned to `SP_IRQ` and `SP_FIQ`
+| Abort  | Always assigned to `SP_ABT`
+| Thread | Assigned to `SP_SVC` while a thread is active
+
+*Notes for Aarch64:*
```
Comment (jbech-linaro) [inline documentation/optee_design.md]: s/Aarch64/AArch64/g

File: documentation/optee_design.md [hunk 4]
```diff
@@ -315,9 +315,105 @@ TA. This is solved by always setting the TA context in the CPU when
...
+pointer.
+
+## Normal entry
+Each time OP-TEE is entered from normal world the temp stack is used
+as the initial stack. For fast calls this is the only stack used. For
+normal calls an empty thread slot is selected and the CPU switches to
+that stack.
+
+## Normal exit
+Normal exit occurs when a thread is finished its task and the thread is
```
Comment (jbech-linaro) [inline documentation/optee_design.md]: "is finished" -> "has finished"

File: documentation/optee_design.md [hunk 5]
```diff
@@ -315,9 +315,105 @@ TA. This is solved by always setting the TA context in the CPU when
...
+that saves the thread state instead. The thread is resumed in the same way
+though.
+
+*Notes for ARMv7/Aarch32:*
+SP_IRQ is initialized to temp stack instead of a separate stack.  Prior to
+exiting to normal world CPU state is changed to SVC and temp stack is
+selected.
+
+*Notes for Aarch64:*
+`SP_EL0` is assigned temp stack and is selected during irq processing. The
```
Comment (jbech-linaro) [inline documentation/optee_design.md]: irq -> IRQ

**Comment by jbech-linaro** [comment]
When comments has been addressed:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: documentation/optee_design.md (hunk 1 above)
Comment (jenswi-linaro) [inline documentation/optee_design.md]: Yes, but it has some drawbacks also. If we say ARMv7-A then we're precise and at the same time excludes ARMv7-M or whatever. Everything we're currently doing we're doing on -A processors, but is everything only applicable for -A processors? 

What should I replace `ARMv7/AArch32` with?

**Comment by jenswi-linaro** [comment]
Rebased and tags applied.

**Pull request merged**